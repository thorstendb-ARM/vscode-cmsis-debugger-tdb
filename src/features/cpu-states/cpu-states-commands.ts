/**
 * Copyright 2025 Arm Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as vscode from 'vscode';
import { EXTENSION_NAME } from '../../manifest';
import { CpuStates } from './cpu-states';

export class CpuStatesCommands {
    public static readonly showCpuTimeHistoryID = `${EXTENSION_NAME}.showCpuTimeHistory`;
    public static readonly resetCpuTimeHistoryID = `${EXTENSION_NAME}.resetCpuTimeHistory`;
    private cpuStates?: CpuStates;

    public activate(context: vscode.ExtensionContext, cpuStates: CpuStates): void {
        // Register item and command
        context.subscriptions.push(
            vscode.commands.registerCommand(CpuStatesCommands.showCpuTimeHistoryID, () => this.handleShowHistory()),
            vscode.commands.registerCommand(CpuStatesCommands.resetCpuTimeHistoryID, () => this.handleResetHistory())
        );
        this.cpuStates = cpuStates;
    }

    protected async handleShowHistory(): Promise<void> {
        if (!this.cpuStates) {
            return;
        }
        await this.cpuStates.updateFrequency();
        this.cpuStates.showStatesHistory();
    }

    protected async handleResetHistory(): Promise<void> {
        if (!this.cpuStates) {
            return;
        }
        this.cpuStates.resetStatesHistory();
    }
};
