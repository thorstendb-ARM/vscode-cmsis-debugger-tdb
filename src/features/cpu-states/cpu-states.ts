/**
 * Copyright 2025 Arm Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as vscode from 'vscode';
import {
    ContinuedEvent,
    GDBTargetDebugTracker,
    SessionStackItem,
    StackTraceRequest,
    StackTraceResponse,
    StoppedEvent
} from '../../debug-session';
import { GDBTargetDebugSession } from '../../debug-session/gdbtarget-debug-session';
import { CpuStatesHistory } from './cpu-states-history';
import { calculateTime, extractPname } from '../../utils';
import { GDBTargetConfiguration } from '../../debug-configuration';
import { DebugProtocol } from '@vscode/debugprotocol';

// Architecturally defined registers (M-profile)
const DWT_CTRL_ADDRESS = 0xE0001000;
const DWT_CYCCNT_ADDRESS = 0xE0001004;

const DWT_CTRL_NOCYCCNT = 0x02000000; // Only for v7-M and v8-M
const DWT_CTRL_CYCCNTENA = 0x00000001;

interface SessionCpuStates {
    states: bigint;
    frequency: number|undefined;
    lastCycles?: number;
    statesHistory: CpuStatesHistory;
    isRunning: boolean;
    hasStates: boolean|undefined;
}

export class CpuStates {
    private readonly _onRefresh: vscode.EventEmitter<number> = new vscode.EventEmitter<number>();
    public readonly onRefresh: vscode.Event<number> = this._onRefresh.event;

    public activeSession: GDBTargetDebugSession | undefined;
    private sessionCpuStates: Map<string, SessionCpuStates> = new Map();
    private stackTraceRequests: Map<string, Map<number, number>> = new Map();

    public get activeCpuStates(): SessionCpuStates|undefined {
        if (!this.activeSession) {
            return undefined;
        }
        return this.sessionCpuStates.get(this.activeSession?.session.id);
    }

    public activate(tracker: GDBTargetDebugTracker): void {
        tracker.onWillStartSession(session => this.handleOnWillStartSession(session));
        tracker.onWillStopSession(session => this.handleOnWillStopSession(session));
        tracker.onDidChangeActiveDebugSession(session => this.handleActiveSessionChanged(session));
        tracker.onDidChangeActiveStackItem(item => this.handleOnDidChangeActiveStackItem(item));
        tracker.onConnected(session => this.handleConnected(session));
        tracker.onContinued(event => this.handleContinuedEvent(event));
        tracker.onStopped(event => this.handleStoppedEvent(event));
        tracker.onStackTraceRequest(request => this.handleStackTraceRequest(request));
        tracker.onStackTraceResponse(response => this.handleStackTraceResponse(response));
    }

    protected handleOnWillStartSession(session: GDBTargetDebugSession): void {
        const states: SessionCpuStates = {
            states: BigInt(0),
            frequency: undefined,
            statesHistory: new CpuStatesHistory(extractPname(session.session.name)),
            isRunning: true,
            hasStates: undefined
        };
        this.sessionCpuStates.set(session.session.id, states);
    }

    protected handleOnWillStopSession(session: GDBTargetDebugSession): void {
        this.stackTraceRequests.delete(session.session.id);
        this.sessionCpuStates.delete(session.session.id);
        if (this.activeSession?.session.id && this.activeSession?.session.id === session.session.id) {
            this.activeSession = undefined;
        }
    }

    protected handleActiveSessionChanged(session?: GDBTargetDebugSession): void {
        this.activeSession = session;
        this._onRefresh.fire(0);
    }

    protected async handleConnected(session: GDBTargetDebugSession): Promise<void> {
        const cpuStates = this.sessionCpuStates.get(session.session.id);
        if (!cpuStates) {
            return;
        }
        // Following call might fail if target not stopped on connect, returns undefined
        // Retry on first Stopped Event.
        cpuStates.hasStates = await this.supportsCpuStates(session);
    }

    protected handleContinuedEvent(event: ContinuedEvent): void {
        const cpuStates = this.sessionCpuStates.get(event.session.session.id);
        if (!cpuStates) {
            return;
        }
        cpuStates.isRunning = true;
    }

    protected async handleStoppedEvent(event: StoppedEvent): Promise<void> {
        const cpuStates = this.sessionCpuStates.get(event.session.session.id);
        if (!cpuStates) {
            return;
        }
        cpuStates.isRunning = false;
        if (cpuStates.hasStates === undefined) {
            // Retry if early read after launch/attach response failed (e.g. if
            // target was running).
            cpuStates.hasStates = await this.supportsCpuStates(event.session);
        }
        if (!cpuStates.hasStates) {
            return;
        }
        return this.updateCpuStates(event.session, event.event.body.threadId, event.event.body.reason);
    }

    protected handleStackTraceRequest(request: StackTraceRequest): void {
        const cpuStates = this.sessionCpuStates.get(request.session.session.id);
        if (!cpuStates) {
            // No need to continue
            return;
        }
        let stackTraceRequests = this.stackTraceRequests.get(request.session.session.id);
        if (stackTraceRequests === undefined) {
            stackTraceRequests = new Map();
            this.stackTraceRequests.set(request.session.session.id, stackTraceRequests);
        }
        stackTraceRequests.set(request.request.seq, request.request.arguments.threadId);
    }

    protected handleStackTraceResponse(response: StackTraceResponse): void {
        // Retrieve and delete tracked request from map first
        const stackTraceRequest = this.stackTraceRequests.get(response.session.session.id);
        const threadId = stackTraceRequest?.get(response.response.request_seq);
        stackTraceRequest?.delete(response.response.request_seq);
        const states = this.activeCpuStates;
        if (!states) {
            return;
        }
        const responseBody: DebugProtocol.StackTraceResponse['body'] = response.response.body;
        const hasValidFrames = response.response.success && responseBody.totalFrames;
        const topFrame = hasValidFrames ? responseBody.stackFrames[0] : undefined;
        let locationString = '';
        if (topFrame) {
            if (topFrame.instructionPointerReference) {
                locationString += `PC=${topFrame.instructionPointerReference} `;
            }
            locationString += `<${topFrame.name}>`;
            if (topFrame.source?.name) {
                locationString += `, ${topFrame.source?.name}`;
            }
            if (topFrame.line) {
                locationString += `::${topFrame.line}`;
            }
        }
        states.statesHistory.insertStopLocation(locationString, threadId);
        // Workaround for VS Code not automatically selecting stack frame without source info.
        // Assuming the correct stack frame is selected within 100ms. To revisit if something
        // can be done without fixed delays and duplicate updates.
        this._onRefresh.fire(100);
    }

    protected handleOnDidChangeActiveStackItem(_item: SessionStackItem): void {
        this._onRefresh.fire(0);
    }

    protected async supportsCpuStates(session: GDBTargetDebugSession): Promise<boolean> {
        if (!(session.session.configuration as GDBTargetConfiguration)?.cmsis) {
            // Only enable feature if launch config contains CMSIS parts.
            return false;
        }
        const dwtCtrlValue = await session.readMemoryU32(DWT_CTRL_ADDRESS);
        // Expect DWT CYCCNT to be enabled by debugger backend. This also covers
        // v6-M where DWT_CTRL_NOCYCCNT is not implemented and always reads '0' while
        // not having a cycle counter.
        return dwtCtrlValue !== undefined && !(dwtCtrlValue & DWT_CTRL_NOCYCCNT) && !!(dwtCtrlValue & DWT_CTRL_CYCCNTENA);
    }

    protected async updateCpuStates(session: GDBTargetDebugSession, threadId?: number, reason?: string): Promise<void> {
        // Update for passed session, not necessarily the active session
        const states = this.sessionCpuStates.get(session.session.id);
        if (!states) {
            return;
        }
        const newCycles = await session.readMemoryU32(DWT_CYCCNT_ADDRESS);
        if (newCycles === undefined) {
            return;
        }
        if (states.lastCycles === undefined) {
            states.lastCycles = newCycles;
        }
        const cycleDiff = newCycles - states.lastCycles;
        const cycleAdd = cycleDiff >= 0 ? cycleDiff : newCycles + Math.pow(2, 32) - states.lastCycles;
        // Caution with types...
        states.lastCycles = newCycles;
        states.states += BigInt(cycleAdd);
        states.statesHistory.updateHistory(states.states, threadId, reason);
    }

    protected async getFrequency(): Promise<number|undefined> {
        const frequencyString = await this.activeSession?.evaluateGlobalExpression('SystemCoreClock');
        if (!frequencyString) {
            return undefined;
        }
        const frequencyValue = parseInt(frequencyString);
        return isNaN(frequencyValue) ? undefined : frequencyValue;
    }

    public async getActivePname(): Promise<string|undefined> {
        if (!this.activeSession) {
            return undefined;
        }
        const sessionName = this.activeSession.session?.name;
        if (!sessionName) {
            return undefined;
        }
        const cbuildRunReader = await this.activeSession.getCbuildRun();
        const pnames = cbuildRunReader?.getPnames();
        const pname = pnames && pnames.length > 1 ? extractPname(sessionName) : undefined;
        return pname;
    }

    public activeHasStates(): boolean|undefined {
        if (this.activeCpuStates?.hasStates === undefined) {
            return undefined;
        }
        return this.activeCpuStates.hasStates;
    }

    public async getActiveTimeString(): Promise<string|undefined> {
        if (!this.activeCpuStates || this.activeHasStates() === undefined ) {
            return undefined;
        }
        const pname = await this.getActivePname();
        if (!this.activeCpuStates.isRunning) {
            // Only update frequency while stopped. User previous otherwise
            // to avoid switching between states and time display.
            await this.updateFrequency();
        }
        const cpuName = pname ? ` ${pname} ` : '';
        if (!this.activeHasStates()) {
            return `${cpuName} N/A`;
        }
        const timeString = this.activeCpuStates.frequency === undefined
            ? `${this.activeCpuStates.states.toString()} states`
            : calculateTime(this.activeCpuStates.states, this.activeCpuStates.frequency);
        return `${cpuName} ${timeString}`;
    }

    public async updateFrequency(): Promise<void> {
        const states = this.activeCpuStates;
        if (!states) {
            return;
        }
        const frequency = await this.getFrequency();
        states.frequency = frequency;
        states.statesHistory.frequency = frequency;
    }

    public showStatesHistory(): void {
        const states = this.activeCpuStates;
        if (!states) {
            return;
        }
        if (!states.hasStates) {
            vscode.window.showWarningMessage('CPU Time commands not available for target');
            return;
        }
        states.statesHistory.showHistory();
    }

    public resetStatesHistory(): void {
        const states = this.activeCpuStates;
        if (!states) {
            return;
        }
        if (!states.hasStates) {
            vscode.window.showWarningMessage('CPU Time commands not available for target');
            return;
        }
        states.statesHistory.resetHistory();
        states.states = BigInt(0);
        this._onRefresh.fire(0);
    }
};
