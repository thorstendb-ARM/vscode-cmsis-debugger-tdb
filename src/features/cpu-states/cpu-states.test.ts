/**
 * Copyright 2025 Arm Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as vscode from 'vscode';
import { debugSessionFactory } from '../../__test__/vscode.factory';
import { GDBTargetConfiguration } from '../../debug-configuration';
import { ContinuedEvent, GDBTargetDebugSession, GDBTargetDebugTracker, SessionStackItem, StackTraceRequest, StackTraceResponse, StoppedEvent } from '../../debug-session';
import { CpuStates } from './cpu-states';
import { DebugProtocol } from '@vscode/debugprotocol';
import { waitForMs } from '../../utils';
import { gdbTargetConfiguration } from '../../debug-configuration/debug-configuration.factory';

const TEST_CBUILD_RUN_FILE = 'test-data/multi-core.cbuild-run.yml'; // Relative to repo root

describe('CpuStates', () => {
    const createContinuedEvent = (session: GDBTargetDebugSession, threadId: number): ContinuedEvent => ({
        session,
        event: {
            body: {
                threadId
            }
        } as unknown as DebugProtocol.ContinuedEvent
    });

    const createStoppedEvent = (session: GDBTargetDebugSession, reason: string, threadId: number): StoppedEvent => ({
        session,
        event: {
            body: {
                reason,
                threadId
            }
        } as unknown as DebugProtocol.StoppedEvent
    });

    const createStackTraceRequest = (session: GDBTargetDebugSession, seq: number, threadId: number): StackTraceRequest => ({
        session,
        request: {
            command: 'stackTrace',
            seq,
            type: 'request',
            arguments: {
                threadId
            }
        }
    });

    const createStackFrame = (): DebugProtocol.StackFrame => ({
        column: 0,
        id: 1,
        line: 2,
        name: 'myframe',
        instructionPointerReference: '0x08000396',
        source: {
            name: 'myfunction'
        }
    });

    const createStackTraceResponse = (session: GDBTargetDebugSession, seq: number, request_seq: number): StackTraceResponse => ({
        session,
        response: {
            command: 'stackTrace',
            type: 'response',
            seq,
            request_seq,
            success: true,
            body: {
                totalFrames: 1,
                stackFrames: [
                    createStackFrame()
                ]
            }
        }
    });

    let debugConfig: GDBTargetConfiguration;
    let cpuStates: CpuStates;
    let tracker: GDBTargetDebugTracker;
    let debugSession: vscode.DebugSession;
    let gdbtargetDebugSession: GDBTargetDebugSession;

    beforeEach(() => {
        debugConfig = gdbTargetConfiguration({ cmsis: { cbuildRunFile: TEST_CBUILD_RUN_FILE } });
        cpuStates = new CpuStates();
        tracker = new GDBTargetDebugTracker();
        debugSession = debugSessionFactory(debugConfig);
        gdbtargetDebugSession = new GDBTargetDebugSession(debugSession);
    });

    describe('session management and connection tests', () => {

        it('activates', () => {
            cpuStates.activate(tracker);
        });

        it('manages session lifecycles correctly', async () => {
            cpuStates.activate(tracker);
            // No active session yet
            expect(cpuStates.activeSession).toBeUndefined();
            // Add session
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onWillStartSession.fire(gdbtargetDebugSession);
            expect(cpuStates.activeSession).toBeUndefined();
            // Activate session
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(gdbtargetDebugSession);
            expect(cpuStates.activeSession?.session.id).toEqual(gdbtargetDebugSession.session.id);
            expect(cpuStates.activeSession?.session.name).toEqual(gdbtargetDebugSession.session.name);
            // Deactivate session
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(undefined);
            expect(cpuStates.activeSession).toBeUndefined();
            // Reactivate session
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(gdbtargetDebugSession);
            expect(cpuStates.activeSession).toBeDefined();
            // Delete session
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onWillStopSession.fire(gdbtargetDebugSession);
            expect(cpuStates.activeSession).toBeUndefined();
        });

        it('adds cpu states object with defaults for new session', () => {
            // Activate and add/switch session
            cpuStates.activate(tracker);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onWillStartSession.fire(gdbtargetDebugSession);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(gdbtargetDebugSession);
            // Check defaults
            expect(cpuStates.activeCpuStates).toBeDefined();
            expect(cpuStates.activeCpuStates?.isRunning).toBeTruthy();
            expect(cpuStates.activeCpuStates?.states).toEqual(BigInt(0));
            expect(cpuStates.activeCpuStates?.frequency).toBeUndefined();
            expect(cpuStates.activeHasStates()).toBeUndefined();
            expect(cpuStates.activeCpuStates?.statesHistory).toBeDefined();
        });

        it('detects cpu states is not supported without cmsis config item', async () => {
            delete debugConfig.cmsis;
            // Activate and add/switch session
            cpuStates.activate(tracker);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onWillStartSession.fire(gdbtargetDebugSession);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(gdbtargetDebugSession);
            // Connected
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onConnected.fire(gdbtargetDebugSession);
            await waitForMs(0);
            expect(cpuStates.activeHasStates()).toEqual(false);
        });

        it.each([
            { info: 'not supported (memory access fails)', value: undefined, expected: false },
            { info: 'not supported (disabled)', value: [ 0x00, 0x00, 0x00, 0x00 ], expected: false },
            { info: 'not supported (not present)', value: [ 0x01, 0x00, 0x00, 0x02 ], expected: false },
            { info: 'supported', value: [ 0x01, 0x00, 0x00, 0x00 ], expected: true },
        ])('detects cpu states is $info', async ({ value, expected }) => {
            if (value === undefined) {
                (debugSession.customRequest as jest.Mock).mockRejectedValueOnce(new Error('test'));
            } else {
                const arrayBuffer = new Uint8Array(value).buffer;
                (debugSession.customRequest as jest.Mock).mockResolvedValueOnce({ address: '0xE0001000', data: arrayBuffer });
            }
            // Activate and add/switch session
            cpuStates.activate(tracker);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onWillStartSession.fire(gdbtargetDebugSession);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(gdbtargetDebugSession);
            // Connected
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onConnected.fire(gdbtargetDebugSession);
            // Let events get processed
            await waitForMs(0);
            expect(cpuStates.activeHasStates()).toEqual(expected);
        });

    });

    describe('tests with established connection and CPU states supported', () => {

        beforeEach(async () => {
            // Activate and add/switch session
            cpuStates.activate(tracker);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onWillStartSession.fire(gdbtargetDebugSession);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(gdbtargetDebugSession);
            (debugSession.customRequest as jest.Mock).mockResolvedValueOnce({
                address: '0xE0001000',
                data:  new Uint8Array([ 0x01, 0x00, 0x00, 0x00 ]).buffer
            });
            // Connected
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onConnected.fire(gdbtargetDebugSession);
            await waitForMs(0);
        });

        it('handles running state correctly', async () => {
            // Considered running after connection
            expect(cpuStates.activeCpuStates?.isRunning).toEqual(true);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onStopped.fire(createStoppedEvent(gdbtargetDebugSession, 'step', 0));
            await waitForMs(0);
            expect(cpuStates.activeCpuStates?.isRunning).toEqual(false);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onContinued.fire(createContinuedEvent(gdbtargetDebugSession, 0));
            await waitForMs(0);
            expect(cpuStates.activeCpuStates?.isRunning).toEqual(true);
        });

        it('captures states for multiple continued/stopped events and shows history', async () => {
            const debugConsoleOutput: string[] = [];
            (vscode.debug.activeDebugConsole.appendLine as jest.Mock).mockImplementation(line => debugConsoleOutput.push(line));
            const stopPoints = 5;
            for (let i = 0; i < stopPoints; i++) {
                (debugSession.customRequest as jest.Mock).mockResolvedValueOnce({
                    address: '0xE0001004',
                    data:  new Uint8Array([ i, 0x00, 0x00, 0x00 ]).buffer
                });
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (tracker as any)._onStopped.fire(createStoppedEvent(gdbtargetDebugSession, 'step', 0));
                await waitForMs(0);
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (tracker as any)._onContinued.fire(createContinuedEvent(gdbtargetDebugSession, 0));
                await waitForMs(0);
            }
            expect(cpuStates.activeCpuStates?.states).toEqual(BigInt(4));
            expect(await cpuStates.getActivePname()).toBeUndefined();
            expect(await cpuStates.getActiveTimeString()).toEqual(' 4 states');
            cpuStates.showStatesHistory();
            expect(debugConsoleOutput).toMatchSnapshot();
        });

        it('captures states with overflow for multiple continued/stopped events', async () => {
            (debugSession.customRequest as jest.Mock)
                .mockResolvedValueOnce({
                    address: '0xE0001004',
                    data:  new Uint8Array([ 0xFE, 0xFF, 0xFF, 0xFF ]).buffer
                })
                .mockResolvedValueOnce({
                    address: '0xE0001004',
                    data:  new Uint8Array([ 0xFF, 0xFF, 0xFF, 0xFF ]).buffer
                })
                .mockResolvedValueOnce({
                    address: '0xE0001004',
                    data:  new Uint8Array([ 0x02, 0x00, 0x00, 0x00 ]).buffer
                });
            for (let i = 0; i < 3; i++) {
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (tracker as any)._onStopped.fire(createStoppedEvent(gdbtargetDebugSession, 'step', 0));
                await waitForMs(0);
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (tracker as any)._onContinued.fire(createContinuedEvent(gdbtargetDebugSession, 0));
                await waitForMs(0);
            }
            expect(cpuStates.activeCpuStates?.states).toEqual(BigInt(4));
            expect(await cpuStates.getActivePname()).toBeUndefined();
            expect(await cpuStates.getActiveTimeString()).toEqual(' 4 states');
        });

        it('captures states for multiple continued/stopped events and resets correctly', async () => {
            for (let i = 0; i < 3; i++) {
                (debugSession.customRequest as jest.Mock).mockResolvedValueOnce({
                    address: '0xE0001004',
                    data:  new Uint8Array([ i*10, 0x00, 0x00, 0x00 ]).buffer
                });
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (tracker as any)._onStopped.fire(createStoppedEvent(gdbtargetDebugSession, 'step', 0));
                await waitForMs(0);
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                (tracker as any)._onContinued.fire(createContinuedEvent(gdbtargetDebugSession, 0));
                await waitForMs(0);
            }
            expect(cpuStates.activeCpuStates?.states).toEqual(BigInt(20));
            expect(await cpuStates.getActiveTimeString()).toEqual(' 20 states');
            cpuStates.resetStatesHistory();
            expect(cpuStates.activeCpuStates?.states).toEqual(BigInt(0));
            expect(await cpuStates.getActiveTimeString()).toEqual(' 0 states');
        });

        it('fires refresh events on active stack item change', async () => {
            const delays: number[] = [];
            const listener = (delay: number) => delays.push(delay);
            const disposable = cpuStates.onRefresh(listener);
            expect(disposable).toBeDefined();
            const sessionStackItem: SessionStackItem = {
                session: gdbtargetDebugSession,
                item: undefined
            };
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveStackItem.fire(sessionStackItem);
            await waitForMs(0);
            expect(delays.length).toEqual(1);
            expect(delays[0]).toEqual(0);
        });

        it('passes undefined frequency to states if getting SystemCoreClock throws', async () => {
            (debugSession.customRequest as jest.Mock).mockRejectedValueOnce(new Error('test error'));
            await cpuStates.updateFrequency();
            expect(cpuStates.activeCpuStates).toBeDefined();
            expect(cpuStates.activeCpuStates?.frequency).toBeUndefined();
            expect(cpuStates.activeCpuStates?.statesHistory.frequency).toBeUndefined();
            expect(await cpuStates.getActiveTimeString()).toEqual(' 0 states');
        });

        it('passes undefined frequency to states if getting SystemCoreClock does not return a number', async () => {
            (debugSession.customRequest as jest.Mock).mockReturnValueOnce({ result: 'not a number' });
            await cpuStates.updateFrequency();
            expect(cpuStates.activeCpuStates).toBeDefined();
            expect(cpuStates.activeCpuStates?.frequency).toBeUndefined();
            expect(cpuStates.activeCpuStates?.statesHistory.frequency).toBeUndefined();
            expect(await cpuStates.getActiveTimeString()).toEqual(' 0 states');
        });

        it('passes valid frequency to states', async () => {
            (debugSession.customRequest as jest.Mock).mockReturnValueOnce({ result: '12000000' });
            await cpuStates.updateFrequency();
            expect(cpuStates.activeCpuStates).toBeDefined();
            expect(cpuStates.activeCpuStates?.frequency).toEqual(12000000);
            expect(cpuStates.activeCpuStates?.statesHistory.frequency).toEqual(12000000);
            expect(await cpuStates.getActiveTimeString()).toEqual(' 0ns');
        });

        it('assigns frame location to captured stop point based on threadId match', async () => {
            const debugConsoleOutput: string[] = [];
            (vscode.debug.activeDebugConsole.appendLine as jest.Mock).mockImplementation(line => debugConsoleOutput.push(line));
            (debugSession.customRequest as jest.Mock).mockResolvedValueOnce({
                address: '0xE0001004',
                data:  new Uint8Array([ 10, 0x00, 0x00, 0x00 ]).buffer
            });
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onStopped.fire(createStoppedEvent(gdbtargetDebugSession, 'step', 1 /*threadId*/));
            await waitForMs(0);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onStackTraceRequest.fire(createStackTraceRequest(gdbtargetDebugSession, 1, 1));
            await waitForMs(0);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onStackTraceResponse.fire(createStackTraceResponse(gdbtargetDebugSession, 4, 1));
            await waitForMs(0);
            cpuStates.showStatesHistory();
            expect(debugConsoleOutput.find(line => line.includes('(PC=0x08000396 <myframe>, myfunction::2)'))).toBeDefined();
        });

        it('does not assign frame location to captured stop point due to threadId mismatch', async () => {
            const debugConsoleOutput: string[] = [];
            (vscode.debug.activeDebugConsole.appendLine as jest.Mock).mockImplementation(line => debugConsoleOutput.push(line));
            (debugSession.customRequest as jest.Mock).mockResolvedValueOnce({
                address: '0xE0001004',
                data:  new Uint8Array([ 10, 0x00, 0x00, 0x00 ]).buffer
            });
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onStopped.fire(createStoppedEvent(gdbtargetDebugSession, 'step', 2 /*threadId*/));
            await waitForMs(0);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onStackTraceRequest.fire(createStackTraceRequest(gdbtargetDebugSession, 1, 1));
            await waitForMs(0);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onStackTraceResponse.fire(createStackTraceResponse(gdbtargetDebugSession, 4, 1));
            await waitForMs(0);
            cpuStates.showStatesHistory();
            expect(debugConsoleOutput.find(line => line.includes('(PC=0x08000396 <myframe>::2)'))).toBeUndefined();
        });

    });

    describe('tests with established connection and CPU states not supported', () => {

        beforeEach(async () => {
            // Activate and add/switch session
            cpuStates.activate(tracker);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onWillStartSession.fire(gdbtargetDebugSession);
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onDidChangeActiveDebugSession.fire(gdbtargetDebugSession);
            (debugSession.customRequest as jest.Mock).mockResolvedValueOnce({
                address: '0xE0001000',
                data:  new Uint8Array([ 0x00, 0x00, 0x00, 0x00 ]).buffer
            });
            // Connected
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (tracker as any)._onConnected.fire(gdbtargetDebugSession);
            await waitForMs(0);
        });

        it('shows warning notification on show history that CPU time is not available', async () => {
            const warningMessageSpy = jest.spyOn(vscode.window, 'showWarningMessage');
            cpuStates.showStatesHistory();
            expect(warningMessageSpy).toHaveBeenCalledWith('CPU Time commands not available for target');
            expect(await cpuStates.getActiveTimeString()).toEqual(' N/A');
        });

        it('shows warning notification on reset history that CPU time is not available', async () => {
            const warningMessageSpy = jest.spyOn(vscode.window, 'showWarningMessage');
            cpuStates.resetStatesHistory();
            expect(warningMessageSpy).toHaveBeenCalledWith('CPU Time commands not available for target');
            expect(await cpuStates.getActiveTimeString()).toEqual(' N/A');
        });

    });

});

