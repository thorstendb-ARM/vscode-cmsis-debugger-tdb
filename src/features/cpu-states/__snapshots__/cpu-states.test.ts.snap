// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CpuStates tests with established connection and CPU states supported captures states for multiple continued/stopped events and shows history 1`] = `
[
  "",
  "ΔT   CPU States   Reason   ",
  "-4   Δ3:      1   step     ",
  "-3   Δ2:      1   step     ",
  "-2   Δ1:      1   step     ",
  "-1   Δ0:      1   step     ",
  " 0            4   step     ",
  "",
]
`;
